src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Core/ExtraTopLevelOperatorsTests.fs

```diff
@@ -0,0 +1,213 @@
+    
+    [<Test>]
+    member this.IEnumerable_T() =        
+        // Legit IE
+        let ie = (dict [|(1,1);(2,4);(3,9)|]) :> IEnumerable<KeyValuePair<_,_>>
+        let enum = ie.GetEnumerator()
+        
+        let testStepping() =
+            CheckThrowsInvalidOperationExn(fun () -> enum.Current |> ignore)
+            Assert.AreEqual(enum.MoveNext(), true)
+            Assert.AreEqual(enum.Current, new KeyValuePair<int,int>(1,1))
+            
+            Assert.AreEqual(enum.MoveNext(), true)
+            Assert.AreEqual(enum.Current, new KeyValuePair<int,int>(2,4))
+            Assert.AreEqual(enum.MoveNext(), true)
+            Assert.AreEqual(enum.Current, new KeyValuePair<int,int>(3,9))
+            Assert.AreEqual(enum.MoveNext(), false)
+            CheckThrowsInvalidOperationExn(fun () -> enum.Current |> ignore)
+    
+        testStepping()
+        enum.Reset()
+        testStepping()
+    
+        // Empty IE
+        let ie = [] |> dict :> IEnumerable  // Note no type args
+        let enum = ie.GetEnumerator()
+        
+        CheckThrowsInvalidOperationExn(fun () -> enum.Current |> ignore)
+        Assert.AreEqual(enum.MoveNext(), false)
+        CheckThrowsInvalidOperationExn(fun () -> enum.Current |> ignore)  
+    
+    
+    [<Test>]
+    member this.IDictionary() =        
+        // Legit ID
+        let id = (dict [|(1,1);(2,4);(3,9)|]) :> IDictionary<_,_> 
+        
+        Assert.IsTrue(id.ContainsKey(1))   
+        Assert.IsFalse(id.ContainsKey(5))  
+        Assert.AreEqual(id.[1], 1)  
+        Assert.AreEqual(id.[3], 9) 
+        Assert.AreEqual(id.Keys,   [| 1; 2; 3|])         
+        Assert.AreEqual(id.Values, [| 1; 4; 9|])
+        
+        CheckThrowsNotSupportedException(fun () -> id.[2] <-88)
+
+        CheckThrowsNotSupportedException(fun () -> id.Add(new KeyValuePair<int,int>(4,16)))
+        Assert.IsTrue(id.TryGetValue(1, ref 1))
+        Assert.IsFalse(id.TryGetValue(100, ref 1))
+        CheckThrowsNotSupportedException(fun () -> id.Remove(1) |> ignore)
+        
+        // Empty ID
+        let id = dict [] :> IDictionary<int, int>   // Note no type args  
+        Assert.IsFalse(id.ContainsKey(5))
+        CheckThrowsKeyNotFoundException(fun () -> id.[1] |> ignore)  
+        Assert.AreEqual(id.Keys,   [| |] )
+        Assert.AreEqual(id.Values, [| |] ) 
+
+    [<Test>]
+    member this.``IReadOnlyDictionary on readOnlyDict``() =
+        let irod = (readOnlyDict [|(1,1);(2,4);(3,9)|]) :> IReadOnlyDictionary<_,_> 
+        
+        Assert.IsTrue(irod.ContainsKey(1))   
+        Assert.IsFalse(irod.ContainsKey(5))  
+        Assert.AreEqual(irod.[1], 1)  
+        Assert.AreEqual(irod.[3], 9) 
+        Assert.AreEqual(irod.Keys,   [| 1; 2; 3|])         
+        Assert.AreEqual(irod.Values, [| 1; 4; 9|])
+        
+        Assert.IsTrue(irod.TryGetValue(1, ref 1))
+        Assert.IsFalse(irod.TryGetValue(100, ref 1))
```

Comment:
As I read it, the test only verifies that the key is found, __not__ that the correct value is returned. That may explain why it passes even though @dsyme is sure that the ref is incorrect.

A better test may be

```F#
let b,i = irod.TryGetValue(2)
Assert.AreEqual(true, b)
Assert.AreEqual(2, i)
```